[AST] Remove ParameterTypeFlags from ParenType and TupleType

hamishknight opened the pull request:
The last clients that relied on stashing parameter type flags on these types are now gone.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on include/swift/AST/Types.h:
```diff
@@ -2123,6 +2115,11 @@ class ParameterTypeFlags {
+/// A type that indicates how parameter flags should be handled in an operation
+/// that requires the conversion into a type that doesn't support them, such as
+/// tuples. They must either be dropped, or be enforced to not be present.
+enum class ParameterFlagHandling { Drop, AssertEmpty };
```
This is unsafe IHMO, be behavior should always be - assert if flags are present,  inside the solver `composeTuple` should only be called after appropriate checks have been performed. We really want to know about places where parameters with non-empty flags are converted into a tuple to figure out why the caller is not checking...

hamishknight replied:
Talked with Pavel offline, we're going to keep this as-is, but with `Drop` renamed to `IgnoreNonEmpty`. It turns out that most callers really do want to drop flags, and making this more convoluted runs the risk of callers just building the tuple themselves without explicitly calling out the fact they are dropping flags.

xedin reviewed (commented):
Looks good to me, I only have one reservation and it's about `ParameterFlagHandling` and letting callers get away with dropping flags.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
GitHub got the wrong timestamps for the force push, correct diff is https://github.com/apple/swift/compare/dab78c634d2adfc350f8f5085fce358040a4d2c8..69f59bbc6c8b2201b64b0100a8a62bffb79775e2

hamishknight commented on lib/APIDigester/ModuleAnalyzerNodes.cpp:
```diff
@@ -1601,8 +1601,9 @@ SwiftDeclCollector::constructTypeNode(Type T, TypeInitInfo Info) {
 
-    auto Input = AnyFunctionType::composeTuple(Fun->getASTContext(),
-                                               Fun->getParams());
+    auto Input =
+        AnyFunctionType::composeTuple(Fun->getASTContext(), Fun->getParams(),
+                                      ParameterFlagHandling::IgnoreNonEmpty);
```
@nkcsgexi This preserves the current behavior of ignoring the parameter flags, but might we want to record parameter flags here for e.g `inout`, variadics, or ownership? AFAIK those would affect ABI. Should I file a bug?

nkcsgexi replied:
Thanks for pinging, @hamishknight ! Wouldn't these flags be reflected in the mangled names of the function? We diagnose mangled name changes as ABI breakage.

hamishknight replied:
Ah yes, in which case I guess this is fine

hamishknight committed:
[CS] Better diagnose inout argument in tuple construction

Previously we relied on `TupleTypeElt::getType`
returning an `InOutType` to fail the tuple type
matching logic. Instead, add logic to reject any
inout arguments up-front with a more specific
diagnostic.

Also, while we're here, strip the `_const`
parameter flag, as it's not something that needs
to be considered for tuple construction.

hamishknight committed:
[AST] Enforce that `composeTuple` drops parameter flags

Callers may either assert that the parameter flags
are empty, or ask for them to be dropped.

hamishknight committed:
[CS] Remove an obsolete bit of logic

SILGen shouldn't care whether or not a self
parameter is wrapped in a paren with ownership.

hamishknight committed:
[AST] Remove ParameterTypeFlags from ParenType and TupleType

The last clients that relied on stashing parameter
type flags on these types are now gone.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@xedin Noticed that we didn't have logic to specifically reject inout arguments for tuple construction (previously we were relying on `TupleTypeElt::getType` producing `InOutType` to fail the match). Added logic to reject inout arguments up-front in 1bf954c61feaed20a7020f91739fc93e49a1c333, mind taking a look?

xedin commented:
@hamishknight Looks great!

hamishknight commented:
:ship:
